### PR TITLE
Feat(Node Embedding): Embedding Node Features

### PR DESCRIPTION
**Pull request by szemyd**
Extracted info necessary for embedding layers, created embedding layers.

### CODE REVIEW

File: data/saved/transactions.csv [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+2018-09-20,002b3c0a44a22c45a8d62ea9d2b88d1a89e335f8b8400335a85c3b5845794cb9,464277014,0.0220169491525423,2
+2018-09-20,003ac6eff3ca20cbe4e14a106a5e28b7551616005de8f9dd0f9455e5543b31ac,617322003,0.0338813559322033,1
+2018-09-20,003ac6eff3ca20cbe4e14a106a5e28b7551616005de8f9dd0f9455e5543b31ac,617322003,0.0186271186440677,1
+2018-09-20,003ac6eff3ca20cbe4e14a106a5e28b7551616005de8f9dd0f9455e5543b31ac,620405001,0.0186271186440677,1
+2018-09-20,003ac6eff3ca20cbe4e14a106a5e28b7551616005de8f9dd0f9455e5543b31ac,678239001,0.0254067796610169,1
+2018-09-20,00401a367c5ac085cb9d4b77c56f3edcabf25153615db982fcc7991d182e10a9,613456009,0.0169322033898305,2
+2018-09-20,00401a367c5ac085cb9d4b77c56f3edcabf25153615db982fcc7991d182e10a9,633675001,0.0101525423728813,2
+2018-09-20,00401a367c5ac085cb9d4b77c56f3edcabf25153615db982fcc7991d182e10a9,648719001,0.0254067796610169,2
+2018-09-20,00401a367c5ac085cb9d4b77c56f3edcabf25153615db982fcc7991d182e10a9,427114015,0.0169322033898305,2
+2018-09-20,00401a367c5ac085cb9d4b77c56f3edcabf25153615db982fcc7991d182e10a9,567475001,0.0338813559322033,2
```
Comment (itchingpixels) [inline data/saved/transactions.csv:67]: I don't think we should commit these csvs, no?

File: run_pipeline.py [hunk 2]
```diff
@@ -20,12 +21,36 @@ def weighted_mse_loss(pred, target, weight=None):
...
+                    int(customer_info.num_cat[i]), int(customer_info.embedding_size[i])
+                )
+            )
+
+        self.embedding = ModuleList(embedding_modules)
         self.conv1 = SAGEConv((-1, -1), hidden_channels)
         self.conv2 = SAGEConv((-1, -1), out_channels)
 
     def forward(self, x, edge_index):
+        embeddings = []
```
Comment (itchingpixels) [inline run_pipeline.py:49]: array comprehension?

File: run_pipeline.py [hunk 3]
```diff
@@ -85,6 +110,27 @@ def test(data, model):
     return float(rmse), model
 
 
+def get_feature_info(full_data: HeteroData) -> tuple[FeatureInfo, FeatureInfo]:
+    customer_features = full_data.x_dict["customer"]
+    article_features = full_data.x_dict["article"]
+
+    customer_num_cat, _ = torch.max(customer_features, dim=0)
+    article_num_cat, _ = torch.max(article_features, dim=0)
```
Comment (itchingpixels) [inline run_pipeline.py:118]: the number of categories is different for each feature, so we'll need to deal identify that per feature (vector dimension)

File: run_preprocessing.py [hunk 4]
```diff
@@ -9,6 +9,14 @@
 from typing import Tuple
 
 
+def save_to_csv(
```
Comment (itchingpixels) [inline run_preprocessing.py:12]: I'm not entirely sure if we want to do this for the whole graph (will take Gbs)? could we make it optional (in DataConfig)?

File: run_pipeline.py (hunk 3 above)
Comment (itchingpixels) [reply run_pipeline.py:118]: thanks for explaining!

File: data/saved/transactions.csv (hunk 1 above)
Comment (szemyd) [reply data/saved/transactions.csv]: Yes, deleted.

File: run_pipeline.py (hunk 2 above)
Comment (szemyd) [reply run_pipeline.py]: This was the missing part.

File: run_preprocessing.py (hunk 4 above)
Comment (szemyd) [reply run_preprocessing.py:12]: Yes, I did it for debugging purposes. I now made it optional.

**Review by itchingpixels — APPROVED** [review]
👍

**Pull request merged**